### PR TITLE
Store points in a MongoDb

### LINKED ISSUE #12: Store points in a database
https://devcenter.heroku.com/articles/mongolab or https://devcenter.heroku.com/articles/ormongo

### PR DESCRIPTION
**Pull request by lucasmontano**
Closes #12

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
...
-  } catch (error) {
-    console.log(error);
-  }
-}, 60 * 1000);
+}
 
-app.get('/', (req, res) => {
-  const factoryViewer = ([name, points]) => ({ name, points });
-  const viewers = Array.from(participants).map(factoryViewer);
+async function incrementParticipantPoints(
```
Comment (EduardoAraujoB) [inline src/app.ts:49]: this function is using the .then and .catch Promises format, so the async/await has no effect on it

File: src/app.ts [hunk 2]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
...
+});
+
+function fetchTopParticipants(participantsCollection: any) {
+  participantsCollection
+    .find()
+    .sort({ points: -1 })
+    .limit(10)
+    .toArray((err, document) => {
+      if (err) throw err;
+      topParticipants = document.map((participant) => {
```
Comment (matheusmelchiades) [inline src/app.ts:43]: depois de adicionado a projection
remover esse trecho e colocar:
```javascript
return document
```

File: src/app.ts [hunk 3]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
...
+  fetchTopParticipants(participantsCollection);
+  setInterval(async () => {
+    try {
+      const viewers = await ChatterService.getViewers();
 
-    viewers.forEach((viewer) => {
-      let point = participants.get(viewer) ?? 0;
-      participants.set(viewer, ++point);
-      console.log(`${viewer} have ${point} points`);
+      viewers.forEach(async (viewer) => {
```
Comment (matheusmelchiades) [inline src/app.ts:25]: melhor performance
```javascript
const promises = viewers.map(viewer => incrementParticipantPoints(viewer))

await Promise.all(promises)
```

File: src/app.ts [hunk 4]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
 const app = express();
 app.use(express.json());
 
-const participants = new Map();
+let topParticipants = new Array();
```
Comment (brunooomelo) [inline src/app.ts:8]: Este new Array não influencia muito.
Acho mais legível 
let topParticipants = []

File: src/app.ts [hunk 5]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
...
+      });
+
+      fetchTopParticipants(participantsCollection);
+    } catch (error) {
+      console.log(error);
+    }
+  }, 60 * 1000);
+});
+
+function fetchTopParticipants(participantsCollection: any) {
```
Comment (d-klotz) [inline src/app.ts:36]: participantsCollection could have its own type instead of 'any'

File: src/app.ts [hunk 6]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
...
 
-app.get('/', (req, res) => {
-  const factoryViewer = ([name, points]) => ({ name, points });
-  const viewers = Array.from(participants).map(factoryViewer);
+async function incrementParticipantPoints(
+  participantsCollection: any,
+  viewer: string
+) {
+  await participantsCollection
+    .updateOne({ name: viewer }, { $inc: { points: 1 } })
```
Comment (omurilo) [inline src/app.ts:54]:
```js
async function incrementParticipantPoints( participantsCollection: Collection,
  viewer: string
): Promise<void> {
  try {
     await participantsCollection.updateOne({ name: viewer }, { $inc: { points: 1 }, upsert: true } });
  } catch (error) {
     console.error(`Failed to add review: ${error}`);
  }
```

Comment (omurilo) [reply src/app.ts:54]: Remove Promise notation and use upsert to insert if not exist on database.

File: src/app.ts [hunk 7]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
...
+
+      fetchTopParticipants(participantsCollection);
+    } catch (error) {
+      console.log(error);
+    }
+  }, 60 * 1000);
+});
+
+function fetchTopParticipants(participantsCollection: any) {
+  participantsCollection
```
Comment (matheusmelchiades) [inline src/app.ts:37]: If you go to use "Promises mode oldSchol"
```javascript
return new Promise((resolve, reject) => {
    participansCollection.find()
          ...restFunctions
          .toArray((err, array) => {
               if (err) return reject(err)
               
                return resolve(array)
           })
})
```

File: src/app.ts [hunk 8]
```diff
@@ -5,30 +5,66 @@ import ChatterService from './services/ChatterService';
...
+const MongoClient = require('mongodb').MongoClient;
+const client = new MongoClient(process.env.MONGODB_URI, {
+  useNewUrlParser: true,
+  useUnifiedTopology: true,
+});
+client.connect((err) => {
+  if (err) throw err;
+
+  const db = client.db(process.env.DATABASE_NAME);
+  const participantsCollection = db.collection('participants');
```
Comment (omurilo) [inline src/app.ts:19]:
```js
interface IParticipants {
    name: string;
    points: number;
}

const participantsCollection: Collection<IParticipants> = db.collection('participants');
```

**Pull request merged**